connector_equipment_service/models/agreement_serviceprofile.py

```diff
@@ -8,69 +8,135 @@
 class AgreementServiceProfile(models.Model):
     _inherit = ['agreement.serviceprofile']
 
-    @api.multi
-    def write(self, vals):
-        # If equipment was empty and now set to managed equipment
-        if vals.get('equipment_id', False) and not self.equipment_id:
+    @api.model
+    def create(self, vals):
+        # If SP is created with a managed equipment
+        if vals.get('equipment_id', False):
             new = self.env['maintenance.equipment'].\
                 browse(vals.get('equipment_id'))
-            new._connect('add_service', serviceprofiles=self)
+            new._connect('add_service',
+                         serviceprofiles=self)
+        return super().create(vals)
+
+    @api.multi
+    def write(self, vals):
+        equip_id = self.get_equip(vals)
+        # Add Service
+        # If equipment was empty and now set to managed or stage in draft
+        if equip_id and (not self.equipment_id or
```

Comment:
This logic can be made simpler (and safer), by first storing the relevant before update values in variables, and then put this logic after the super().write.

In fact, checking for empty fields in `vals` is tricky, because the write operation might not include all fields.
For example, you could `write({'state': 'done'}`, and when checking for `vals.get('equipment_id')` you would get `None`.
But in fact the Equipment value might be set and has not changed at all!

Reply:
@dreispt But if equipment does not exist in vals then we know the equipment is not changed. If the equipment is in vals but it is False then we know they are removing the equipment. Lastly, if equipment is in vals and returns an id then we know they are switching equipment. I believe the code accounts for all of these scenarios.

I think we should jump on a hangout and try to resolved this one

Reply:
Sure, it can work perfectly. But the logic conditions are a bit complex and it's easy to get some cases wrong.
My point is that there is an alternative that could make make the logic simpler.
It we confirm it's working good, no reason to change the code.

I would be more comfortable with a few tests, though:
If this piece of code does not work as expected, it will have a visible impact of the ISP customer service. That could be done in the scope of a different task.